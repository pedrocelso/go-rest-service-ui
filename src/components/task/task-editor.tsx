import React, { Component } from 'react';
import { Grid, Typography, LinearProgress, ExpansionPanel, ExpansionPanelSummary, ExpansionPanelDetails, Button, Dialog, AppBar, Toolbar, IconButton, List, ListItem, ListItemText, Divider, Slide } from '@material-ui/core';
import CloseIcon from '@material-ui/icons/Close';

function Transition(props: any) {
  return <Slide direction="up" {...props} />;
}

export interface TaskEditorProps {
  open: boolean
  close: (b: boolean) => void
}

export class TaskEditor extends Component<TaskEditorProps,{}> {
  handleClose = () => {
    this.props.close(false)
  };

  render() {
    return (
      <Dialog
        fullScreen
        open={this.props.open}
        onClose={this.handleClose}
        TransitionComponent={Transition}
      >
        <AppBar>
          <Toolbar>
            <IconButton color="inherit" onClick={this.handleClose} aria-label="Close">
              <CloseIcon />
            </IconButton>
            <Typography variant="h6" color="inherit" >
              Sound
            </Typography>
            <Button color="inherit" onClick={this.handleClose}>
              save
            </Button>
          </Toolbar>
        </AppBar>
        <List>
          <ListItem button>
            <ListItemText primary="Phone ringtone" secondary="Titania" />
          </ListItem>
          <Divider />
          <ListItem button>
            <ListItemText primary="Default notification ringtone" secondary="Tethys" />
          </ListItem>
        </List>
      </Dialog>
    )
  }
}
