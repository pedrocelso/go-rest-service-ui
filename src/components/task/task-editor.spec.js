import React from 'react';
import { shallow } from 'enzyme'
import sinon from 'sinon'

import { TaskEditor, Transition } from './task-editor'

const baseProps = {
  deauthenticate: sinon.fake(),
  open: sinon.fake(),
  close: sinon.fake(),
  editor: {
    open: false
  },
  service: {
    createTask: (_) => ({
      fork: (rej, res) => {
        rej({ statusCode: 401 })
      }
    })
  }
}

describe(`<TaskEditor />`, () => {
  it(`Should render when it is open`, () => {
    const component = shallow(
      <TaskEditor {...baseProps} />,
    );

    expect(component).toMatchSnapshot();
  });

  it(`Should not render when it is open`, () => {
    const props = {
      ...baseProps,
      editor: {
        open: false
      }
    }
    const component = shallow(
      <TaskEditor {...props} />,
    );

    expect(component).toMatchSnapshot();
  });

  it(`.handleSubmit() - Successful response`, () => {
    const deauthenticate = sinon.spy()
    const service = {
      createTask: (_) => ({
        fork: (rej, res) => {
          res({ task: { id: 10, name: `test`, description: `test@test.com` } })
        }
      })
    }
    const close = sinon.spy()

    const props = {
      ...baseProps,
      deauthenticate,
      close,
      service
    }
    const component = shallow(
      <TaskEditor {...props} />,
    );

    component.setState({ formData: { name: 'bar', description: 'foo' } });
    component.instance().handleSubmit()

    expect(close.called).toBeTruthy()
    expect(component.state().loading).toBeFalsy()
    expect(component.state().formData).toMatchObject({})
  });

  it(`.handleSubmit() - 401 response`, () => {
    const deauthenticate = sinon.spy()
    const service = {
      createTask: (_) => ({
        fork: (rej, res) => {
          rej({ statusCode: 401 })
        }
      })
    }
    const close = sinon.spy()

    const props = {
      ...baseProps,
      deauthenticate,
      close,
      service
    }
    const component = shallow(
      <TaskEditor {...props} />,
    );
    component.setState({ formData: { name: 'bar', description: 'foo' } });
    component.instance().handleSubmit()

    expect(close.called).toBeFalsy()
    expect(deauthenticate.called).toBeTruthy()
    expect(component.state().loading).toBeFalsy()
  });

  it(`.handleSubmit() - Error response`, () => {
    const deauthenticate = sinon.spy()
    const service = {
      createTask: (_) => ({
        fork: (rej, res) => {
          rej(`error fetching tasks`)
        }
      })
    }
    const close = sinon.spy()

    const props = {
      ...baseProps,
      deauthenticate,
      close,
      service
    }
    const component = shallow(
      <TaskEditor {...props} />,
    );

    component.setState({ formData: { name: 'bar', description: 'foo' } });
    component.instance().handleSubmit()
    expect(close.called).toBeFalsy()
    expect(deauthenticate.called).toBeFalsy()
    expect(component.state().loading).toBeFalsy()
    expect(component.state().formData).toMatchObject({ name: 'bar', description: 'foo' })
  });

  it(`.handleClose() - Should close the dialog`, () => {
    const close = sinon.spy()

    const props = {
      ...baseProps,
      close
    }
    const component = shallow(
      <TaskEditor {...props} />,
    );

    component.instance().handleClose()
    expect(close.called).toBeTruthy()
  });

  it(`.handleChange() - Should update the formData`, () => {
    const component = shallow(
      <TaskEditor {...baseProps} />,
    );

    component.instance().handleChange({ formData: { name: `migeh`, description: `maneh` } })
    expect(component.state().formData).toMatchObject({ name: 'migeh', description: 'maneh' })
  });

  it(`Transition - Should render when it is open`, () => {
    const props = {
      ...baseProps,
      editor: {
        open: true
      }
    }
    const component = shallow(
      <TaskEditor {...props} />,
    );

    expect(component).toMatchSnapshot();  
  });
});
