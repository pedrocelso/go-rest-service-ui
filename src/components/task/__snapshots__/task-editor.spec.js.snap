// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskEditor /> Should not render when it is open 1`] = `
<WithStyles(ForwardRef(Dialog))
  TransitionComponent={[Function]}
  className="editor"
  onClose={[Function]}
  open={false}
  scroll="paper"
>
  <WithStyles(ForwardRef(DialogContent))
    className="editor__content"
  >
    <WithStyles(ForwardRef(AppBar))
      position="static"
    >
      <WithStyles(ForwardRef(Toolbar))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Close"
          color="inherit"
          onClick={[Function]}
        >
          <CloseIcon />
        </WithStyles(ForwardRef(IconButton))>
        <WithStyles(ForwardRef(Typography))
          className="toolbar-title"
          color="inherit"
          variant="h6"
        >
          Create
           Task
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Button))
          color="inherit"
          onClick={[Function]}
        >
          Save
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Toolbar))>
    </WithStyles(ForwardRef(AppBar))>
    <div
      className="loading-bar"
    />
    <Form
      ErrorList={[Function]}
      className="editor__form"
      disabled={false}
      formData={Object {}}
      liveValidate={false}
      noHtml5Validate={false}
      noValidate={false}
      omitExtraData={false}
      onChange={[Function]}
      onError={[Function]}
      onSubmit={[Function]}
      safeRenderCompletion={false}
      schema={
        Object {
          "properties": Object {
            "description": Object {
              "default": "",
              "title": "Description",
              "type": "string",
            },
            "name": Object {
              "default": "",
              "title": "Task name",
              "type": "string",
            },
          },
          "required": Array [
            "name",
          ],
          "type": "object",
        }
      }
      uiSchema={
        Object {
          "description": Object {
            "ui:options": Object {
              "rows": 5,
            },
            "ui:widget": "textarea",
          },
        }
      }
    >
      <button
        className="hidden"
      />
    </Form>
  </WithStyles(ForwardRef(DialogContent))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`<TaskEditor /> Should render when it is open 1`] = `
<WithStyles(ForwardRef(Dialog))
  TransitionComponent={[Function]}
  className="editor"
  onClose={[Function]}
  open={false}
  scroll="paper"
>
  <WithStyles(ForwardRef(DialogContent))
    className="editor__content"
  >
    <WithStyles(ForwardRef(AppBar))
      position="static"
    >
      <WithStyles(ForwardRef(Toolbar))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Close"
          color="inherit"
          onClick={[Function]}
        >
          <CloseIcon />
        </WithStyles(ForwardRef(IconButton))>
        <WithStyles(ForwardRef(Typography))
          className="toolbar-title"
          color="inherit"
          variant="h6"
        >
          Create
           Task
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Button))
          color="inherit"
          onClick={[Function]}
        >
          Save
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Toolbar))>
    </WithStyles(ForwardRef(AppBar))>
    <div
      className="loading-bar"
    />
    <Form
      ErrorList={[Function]}
      className="editor__form"
      disabled={false}
      formData={Object {}}
      liveValidate={false}
      noHtml5Validate={false}
      noValidate={false}
      omitExtraData={false}
      onChange={[Function]}
      onError={[Function]}
      onSubmit={[Function]}
      safeRenderCompletion={false}
      schema={
        Object {
          "properties": Object {
            "description": Object {
              "default": "",
              "title": "Description",
              "type": "string",
            },
            "name": Object {
              "default": "",
              "title": "Task name",
              "type": "string",
            },
          },
          "required": Array [
            "name",
          ],
          "type": "object",
        }
      }
      uiSchema={
        Object {
          "description": Object {
            "ui:options": Object {
              "rows": 5,
            },
            "ui:widget": "textarea",
          },
        }
      }
    >
      <button
        className="hidden"
      />
    </Form>
  </WithStyles(ForwardRef(DialogContent))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`<TaskEditor /> Transition - Should render when it is open 1`] = `
<WithStyles(ForwardRef(Dialog))
  TransitionComponent={[Function]}
  className="editor"
  onClose={[Function]}
  open={true}
  scroll="paper"
>
  <WithStyles(ForwardRef(DialogContent))
    className="editor__content"
  >
    <WithStyles(ForwardRef(AppBar))
      position="static"
    >
      <WithStyles(ForwardRef(Toolbar))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="Close"
          color="inherit"
          onClick={[Function]}
        >
          <CloseIcon />
        </WithStyles(ForwardRef(IconButton))>
        <WithStyles(ForwardRef(Typography))
          className="toolbar-title"
          color="inherit"
          variant="h6"
        >
          Create
           Task
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Button))
          color="inherit"
          onClick={[Function]}
        >
          Save
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Toolbar))>
    </WithStyles(ForwardRef(AppBar))>
    <div
      className="loading-bar"
    />
    <Form
      ErrorList={[Function]}
      className="editor__form"
      disabled={false}
      formData={Object {}}
      liveValidate={false}
      noHtml5Validate={false}
      noValidate={false}
      omitExtraData={false}
      onChange={[Function]}
      onError={[Function]}
      onSubmit={[Function]}
      safeRenderCompletion={false}
      schema={
        Object {
          "properties": Object {
            "description": Object {
              "default": "",
              "title": "Description",
              "type": "string",
            },
            "name": Object {
              "default": "",
              "title": "Task name",
              "type": "string",
            },
          },
          "required": Array [
            "name",
          ],
          "type": "object",
        }
      }
      uiSchema={
        Object {
          "description": Object {
            "ui:options": Object {
              "rows": 5,
            },
            "ui:widget": "textarea",
          },
        }
      }
    >
      <button
        className="hidden"
      />
    </Form>
  </WithStyles(ForwardRef(DialogContent))>
</WithStyles(ForwardRef(Dialog))>
`;
